gtsfm/averaging/rotation/shonan.py

```diff
@@ -30,20 +30,33 @@ class ShonanRotationAveraging(RotationAveragingBase):
     """Performs Shonan rotation averaging."""
 
     def __init__(self) -> None:
+        """
+        Note: `p_min` and `p_max` describe the minimum and maximum relaxation rank.
+        """
         self._p_min = 5
         self._p_max = 30
 
-    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
-        """Run the rotation averaging.
+    def run_w_consecutive_ordering(
+        self, num_connected_nodes: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
+    ) -> List[Optional[Rot3]]:
+        """Run the rotation averaging on a connected graph w/ N keys ordered consecutively [0,...,N-1].
+
+        Note: the assumption that Shonan keys are ordered as [0,...,N-1] goes deeper than
+        ShonanAveraging::NrUnknowns(), which just checks the assumption. Several other functions that construct
+        matrices related to the convergence certificate also use the fact that the graph is connected and
+        keys are 0..N-1. There are about 10 places where nrUnknowns is used to dimension a sparse or dense matrix.
+        Modifying GTSAM would require a major philosophical overhaul, so we perform the re-ordering
+        here in a sort of "wrapper". See: https://github.com/borglab/gtsam/issues/784
 
         Args:
-            num_images: number of poses.
-            i2Ri1_dict: relative rotations as dictionary (i1, i2): i2Ri1.
+            num_connected_nodes: number of unique connected nodes in the graph (can be far less than the number
+                of images in the dataset)
+            i2Ri1_dict: relative rotations for each image pair-edge as dictionary (i1, i2): i2Ri1.
 
         Returns:
-            Global rotations for each camera pose, i.e. wRi, as a list. The number of entries in the list is
-                `num_images`. The list may contain `None` where the global rotation could not be computed (either
-                underconstrained system or ill-constrained system).
+            Global rotations for each **CONNECTED** camera pose, i.e. wRi, as a list. The number of entries in
+                the list is num_connected_nodes`. The list may contain `None` where the global rotation could
+                not be computed (either underconstrained system or ill-constrained system).
```

Comment:
If the input edges belong to a connected component, it seems like this function returns wRi for all input nodes, except when the system is underconstrained.

Reply:
Hmm didn't quite follow this here -- so the distinction is about underconstrained vs. ill-constrained?

Reply:
I dont know the difference... what is it? 

I did not know num_images was num_connected_nodes. I thought it was num_nodes (>= num_connected_nodes). Hence the confusion.